types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+		struct wl_resource *resource) {
+	assert(
+		wl_resource_instance_of(
+			resource, &zwp_confined_pointer_v1_interface,
+				&confined_pointer_impl) ||
+		wl_resource_instance_of(
+			resource, &zwp_locked_pointer_v1_interface,
+				&locked_pointer_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_pointer_constraints_v1 *pointer_constraints_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_pointer_constraints_v1_interface,
+		&pointer_constraints_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+// Thankfully zwp_confined_pointer_v1_send_{un,}confined work on both locked and confined pointer constraints
+static void send_constrained(struct wlr_pointer_constraint_v1 *constraint) {
+	constraint->active = true;
+
+	wlr_log(L_DEBUG, "constrained %p", constraint);
+	zwp_confined_pointer_v1_send_confined(constraint->resource);
+}
+static void send_unconstrained(struct wlr_pointer_constraint_v1 *constraint) {
+	if (!constraint->active) {
+		return;
+	}
+	constraint->active = false;
+
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_inactive,
+		constraint);
+
+	wlr_log(L_DEBUG, "unconstrained %p", constraint);
+	zwp_confined_pointer_v1_send_unconfined(constraint->resource);
+	if (constraint->lifetime == ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT) {
+		wl_resource_destroy(constraint->resource);
+	}
+}
+
+static void pointer_constraint_check_active(struct wlr_pointer_constraint_v1 *constraint) {
+	struct wlr_pointer_constraint_v1 *current_constraint =
+		constraint->seat_client->seat->current_pointer_constraint;
+	if (current_constraint != constraint ||
+			constraint->current.region || !constraint->accepted) {
+		return;
+	}
+
+	struct wlr_pointer_constraint_v1_activation activation = {
+		.pointer_constraint = constraint,
+		.box = NULL,
+	};
+
+	send_constrained(constraint);
+	wlr_signal_emit_safe(
+		&constraint->pointer_constraints->events.constraint_active,
+		&activation);
+}
+
+void wlr_pointer_constraint_v1_accept(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p accepted", constraint);
+	if (constraint->accepted) {
+		return;
+	}
+	constraint->accepted = true;
+	pointer_constraint_check_active(constraint);
+}
+
+void wlr_pointer_constraint_v1_decline(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p declined", constraint);
+	constraint->accepted = false;
+	send_unconstrained(constraint);
+}
+
+static void resource_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint) {
+	if (constraint == NULL) {
+		return;
+	}
+
+	if (constraint->seat_client->seat->current_pointer_constraint == constraint) {
+		constraint->seat_client->seat->current_pointer_constraint = NULL;
```

Comment:
We normally do this by emitting a destroy signal and having everything that takes a reference do its own cleanup.